### PR TITLE
Add db search subcommand

### PR DESCRIPTION
**Pull request by tomersein**
Sometimes I want to understand if the DB has the vulnerability I am looking for.
Today in order to do that I need to go to listing.json -> get the latest URL -> Download it -> open DBeaver -> use a simple SQL query.

This PR creates a new endpoint which can accomplish it in a simple command grype explore cve <cve_id>

solves - https://github.com/anchore/grype/issues/1629

### CODE REVIEW

File: cmd/grype/cli/cli.go [hunk 1]
```diff
@@ -81,6 +81,7 @@ func create(id clio.Identification) (clio.Application, *cobra.Command) {
 		commands.DB(app),
 		commands.Completion(app),
 		commands.Explain(app),
+		commands.Explore(app),
```
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/cli.go:84]: can do so, I wanted to make the less changes possible in the current functions

File: cmd/grype/cli/commands/explore_cve.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		Short: "explore a vulnerability and display information",
+		Args:  cobra.MaximumNArgs(1),
+		RunE: func(_ *cobra.Command, args []string) (err error) {
+			switch len(args) {
+			case 1:
+				return runExploreCVE(opts, args[0])
+			default:
+				return fmt.Errorf("requires a single vulnerability ID'")
+			}
+		},
```
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/commands/explore_cve.go]: please look at my change :)

File: cmd/grype/cli/cli.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/cli.go]: look at my change :)

File: cmd/grype/cli/commands/db_query.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func ExploreCVE(app clio.Application) *cobra.Command {
+	opts := &dbQueryOptions{
+		Output:    "table",
+		DBOptions: *dbOptionsDefault(app.ID()),
+	}
+
+	return app.SetupCommand(&cobra.Command{
+		Use:   "query [flags] vulnerability_id",
```
Comment (kzantow) [inline cmd/grype/cli/commands/db_query.go:41]: this no longer matches the implementation, it looks like. you can omit the `[flags]`, these are printed below when you run `--help` no matter what. I know some other commands have this, but it's not necessary or typical across most of our CLIs

File: cmd/grype/cli/commands/db_query.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		Short: "query the db and display information",
+		Args:  cobra.MaximumNArgs(1),
```
[the remark this answers is not in the document]
Comment (kzantow) [reply cmd/grype/cli/commands/db_query.go:43]: I thought there were more asks for this type of thing, but this was the only other one I could find: https://github.com/anchore/grype/issues/656, and I guess it doesn't _exactly_ fit into the same type of vulnerability query. I'm going to add this to our weekly gardening session to get the team's input, so we should have a 👍 or direction soon.

File: cmd/grype/cli/commands/db_query.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	Output    string `yaml:"output" json:"output" mapstructure:"output"`
+	DBOptions `yaml:",inline" mapstructure:",squash"`
+	ID        string `yaml:"id" json:"id" mapstructure:"id"`
+}
+
+var _ clio.FlagAdder = (*dbQueryOptions)(nil)
+
+func (c *dbQueryOptions) AddFlags(flags clio.FlagSet) {
+	flags.StringVarP(&c.Output, "output", "o", "format to display results (available=[table, json])")
+	flags.StringVarP(&c.ID, "id", "i", "get information on vulnerability id")
```
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/commands/db_query.go]: fixed

File: cmd/grype/cli/commands/db_query.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/commands/db_query.go:43]: updated

File: cmd/grype/cli/commands/db_query.go (hunk 3 above)
Comment (tomersein) [reply cmd/grype/cli/commands/db_query.go]: fixed

**Comment by popey** [comment]
👋 We will discuss this topic at our next [Open Source Gardening Live Stream](https://www.youtube.com/live/5CbKCpCRdAg) later today. Anyone interested in the topic is welcome to join. All the details are in (this thread](https://anchorecommunity.discourse.group/t/august-8th-open-source-gardening-live-stream/60) 🎉

**Comment by spiffcs** [comment]
If you pull down this branch the easiest way to get this running is the following example:
```
$go run cmd/grype/main.go db query CVE-2024-20932

ID              PACKAGE NAME              NAMESPACE                             VERSION CONSTRAINT
CVE-2024-20932  graalvm_enterprise        nvd:cpe                               >= 21-ea, <= 21.3.8 || >= 22-ea, <= 22.3.4 (unknown)
CVE-2024-20932  graalvm_for_jdk           nvd:cpe                               >= 17, <= 17.0.9 (unknown)
CVE-2024-20932  java_se                   nvd:cpe                               >= 17, <= 17.0.9 (unknown)
CVE-2024-20932  jdk                       nvd:cpe                               >= 17, <= 17.0.9 (unknown)
CVE-2024-20932  jre                       nvd:cpe                               >= 17, <= 17.0.9 (unknown)
CVE-2024-20932  openjdk                   nvd:cpe                               >= 17, <= 17.0.9 (unknown)
CVE-2024-20932  openjdk17                 alpine:distro:alpine:3.16             < 17.0.10_p7-r0 (apk)
CVE-2024-20932  openjdk17                 alpine:distro:alpine:3.17             < 17.0.10_p7-r0 (apk)
CVE-2024-20932  openjdk17                 alpine:distro:alpine:3.18             < 17.0.10_p7-r0 (apk)
CVE-2024-20932  openjdk17                 alpine:distro:alpine:3.19             < 17.0.10_p7-r0 (apk)
CVE-2024-20932  openjdk17                 alpine:distro:alpine:3.20             < 17.0.10_p7-r0 (apk)
CVE-2024-20932  openjdk17                 alpine:distro:alpine:edge             < 17.0.10_p7-r0 (apk)
CVE-2024-20932  openjdk-10                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-11                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-12                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-13                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-14                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-15                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-16                chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-7                 chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-8                 chainguard:distro:chainguard:rolling  < 0 (apk)
CVE-2024-20932  openjdk-9                 chainguard:distro:chainguard:rolling  < 0 (apk)
.......
```

Is query the correct word?

Query to me is usually associated with `sql` or a formal domain specific language.

What do people think about just basic `grype db get <vulnerability_id>` and just remove all the current flags?

**Comment by tomersein** [comment]
I've changed the terminology, removed the id flag.
`db get CVE-2024-6387`

returns:
```
 ✔ Vulnerability DB                [no update available]  
ID             PACKAGE NAME                              NAMESPACE                      VERSION CONSTRAINT           
CVE-2024-6387  openssh                                   nvd:cpe                        >= 8.5p1, < 9.8 (unknown)     
CVE-2024-6387  openssh                                   alpine:distro:alpine:3.17      < 9.1_p1-r6 (apk)             
CVE-2024-6387  openssh                                   alpine:distro:alpine:3.18      < 9.3_p2-r2 (apk)             
CVE-2024-6387  openssh                                   alpine:distro:alpine:3.19      < 9.6_p1-r1 (apk)             
CVE-2024-6387  openssh                                   alpine:distro:alpine:3.20      < 9.7_p1-r4 (apk)             
CVE-2024-6387  openssh                                   alpine:distro:alpine:edge      < 9.8_p1-r0 (apk)             
CVE-2024-6387  openssh                                   debian:distro:debian:12        < 1:9.2p1-2+deb12u3 (deb)     
CVE-2024-6387  openssh                                   debian:distro:debian:13        < 1:9.7p1-7 (deb)             
CVE-2024-6387  openssh                                   debian:distro:debian:unstable  < 1:9.7p1-7 (deb)             
CVE-2024-6387  openssh                                   mariner:distro:mariner:2.0     < 0:8.9p1-6.cm2 (rpm)         
CVE-2024-6387  openssh                                   redhat:distro:redhat:9         < 0:8.7p1-38.el9_4.1 (rpm)    
CVE-2024-6387  openssh                                   sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-clients                           sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-common                            sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-fips                              sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-helpers                           sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-server                            sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-server-config-disallow-rootlogin  sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh-askpass-gnome                     sles:distro:sles:15.6          < 0:9.6p1-150600.6.3.1 (rpm)  
CVE-2024-6387  openssh                                   ubuntu:distro:ubuntu:22.04     < 1:8.9p1-3ubuntu0.10 (deb)   
CVE-2024-6387  openssh                                   ubuntu:distro:ubuntu:23.10     < 1:9.3p1-1ubuntu3.6 (deb)    
CVE-2024-6387  openssh                                   ubuntu:distro:ubuntu:24.04     < 1:9.6p1-3ubuntu13.3 (deb)
```

File: cmd/grype/cli/commands/db_get.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		Use:   "get vulnerability_id",
+		Short: "get information on a vulnerability from the db",
+		Args:  cobra.MaximumNArgs(1),
+		RunE: func(_ *cobra.Command, args []string) (err error) {
+			id := args[0]
+			if id == "" {
+				return fmt.Errorf("specify the vulnerability ID")
+			}
+			return runGetCVE(opts, id)
+		},
```
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/commands/db_get.go]: fixed

File: grype/db/vulnerability_provider.go [hunk 7]
```diff
@@ -51,6 +51,9 @@ func (pr *VulnerabilityProvider) Get(id, namespace string) ([]vulnerability.Vuln
 	var results []vulnerability.Vulnerability
 	for _, vuln := range vulns {
 		vulnObj, err := vulnerability.NewVulnerability(vuln)
+		if len(namespace) == 0 {
+			vulnObj.PackageName = vuln.PackageName
+		}
```
[the remark this answers is not in the document]
Comment (tomersein) [reply grype/db/vulnerability_provider.go:56]: didn't want to change previous behave
up to your decision

File: cmd/grype/cli/commands/db_get.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomersein) [reply cmd/grype/cli/commands/db_get.go]: fixed all comments

File: grype/db/vulnerability_provider.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomersein) [reply grype/db/vulnerability_provider.go]: fixed

File: cmd/grype/cli/commands/db_get.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		err := present(opts.Output, vulnerabilities, sb)
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		}
+	}
+
+	bus.Report(sb.String())
+
+	return errs
+}
```
Comment (kzantow) [inline cmd/grype/cli/commands/db_get.go:88]: Sorry, I should have taken a closer look before. 🤦  This function could be simplified a bit: it doesn't need a single-entry parallel call, everything needs to happen in serial anyway, which simplifies the error handling to not need any multierror append, and the terminology should be updated to match search/vunlerability id. Something like this:

```
func runDbSearch(opts *dbQueryOptions, vulnerabilityID string) error {
	log.Debug("loading DB")
	str, status, dbCloser, err := grype.LoadVulnerabilityDB(opts.DB.ToCuratorConfig(), opts.DB.AutoUpdate)
	err = validateDBLoad(err, status)
	if err != nil {
		return err
	}
	if dbCloser != nil {
		defer dbCloser.Close()
	}

	vulnerabilities, err := str.Get(vulnerabilityID, "")
	if err != nil {
		return err
	}

	sb := &strings.Builder{}
	if len(vulnerabilities) == 0 {
		return fmt.Errorf("Vulnerability doesn't exist in the DB: %s", vulnerabilityID)
	}
	
	err = present(opts.Output, vulnerabilities, sb)
	bus.Report(sb.String())

	return err
}
```

File: grype/db/v5/store/store.go [hunk 9]
```diff
@@ -99,7 +99,13 @@ func (s *store) GetVulnerabilityNamespaces() ([]string, error) {
...
 	var models []model.VulnerabilityModel
 
-	result := s.db.Where("namespace = ? AND id = ?", namespace, id).Find(&models)
+	var query = s.db
+
+	if namespace != "" {
+		query = query.Where("namespace = ?", namespace)
+	}
+
+	result := query.Where("id = ?", id).Find(&models)
```
Comment (kzantow) [inline grype/db/v5/store/store.go:108]: nit: this query builder might be slightly more clear if the ID clause is added on line 102, e.g.:
```
	query := s.db.Where("id = ?", id)

	if namespace != "" {
		query = query.Where("namespace = ?", namespace)
	}

	result := query.Find(&models)
```

**Review by kzantow — COMMENTED** [review]
Sorry for the back-and-forth, I think this is really the last details

File: grype/db/v5/store/store.go (hunk 9 above)
Comment (tomerse-sg) [reply grype/db/v5/store/store.go]: fixed

File: cmd/grype/cli/commands/db_get.go (hunk 8 above)
Comment (tomerse-sg) [reply cmd/grype/cli/commands/db_get.go]: fixed

Comment (tomersein) [reply cmd/grype/cli/commands/db_get.go]: fixed

File: grype/db/v5/store/store.go (hunk 9 above)
Comment (tomersein) [reply grype/db/v5/store/store.go]: fixed

**Review by kzantow — APPROVED** [review]
Thanks for this @tomersein

**Comment by kzantow** [comment]
@tomersein it looks like you might just need to run `make lint-fix` for the static analysis thing, and there seems to be an issue with a failing test -- shout if you want me to push any changes for it

**Comment by tomersein** [comment]
> @tomersein it looks like you might just need to make [a couple fixes here](https://github.com/anchore/grype/actions/runs/10357800009/job/28672034422?pr=2031#step:4:69) for the static analysis thing, and there seems to be an issue with a failing test -- shout if you want me to push any changes for it

fixed the lint & the UT

File: grype/db/vulnerability_provider.go [hunk 10]
```diff
@@ -51,6 +51,7 @@ func (pr *VulnerabilityProvider) Get(id, namespace string) ([]vulnerability.Vuln
 	var results []vulnerability.Vulnerability
 	for _, vuln := range vulns {
 		vulnObj, err := vulnerability.NewVulnerability(vuln)
+		vulnObj.PackageName = vuln.PackageName
```
Comment (kzantow) [inline grype/db/vulnerability_provider.go:54]: I think this just needs to be moved after the error check

Comment (tomersein) [reply grype/db/vulnerability_provider.go]: fixed

**Comment by kzantow** [comment]
Thanks for the contribution, @tomersein (and sorry for the back-and-forth)

**Pull request merged**